### PR TITLE
Chore: Permissions check per endpoint/method

### PR DESCRIPTION
**Pull request by KevLehman**
Note: most files changed are conversions to use the implemented permissions param :eyes: so don't worry to see 40 files.

The idea:
- There are some routes that have multiple handlers per HTTP method. What happens in those routes is that we need to check for permissions on each one of the handlers, which is a bit of boilerplate, and increases the number of code branches. With this change, endpoints will be able to say which permissions apply to each one of the methods, as well as the "operation" that they want the permissions to be.
- For example, an endpoint could be accessed by any user with 'view-l-room' or 'view-m-room'. With the current implementation, that cannot be done. With the new changes, devs can pass a `operation: hasAll | hasAny` prop to signal how permissions should be evaluated, if inclusively or exclusively.

:)

### CODE REVIEW

File: apps/meteor/app/livechat/imports/server/rest/rooms.js [hunk 1]
```diff
@@ -19,7 +19,7 @@ const validateDateParams = (property, date) => {
 
 API.v1.addRoute(
 	'livechat/rooms',
-	{ authRequired: true },
+	{ authRequired: true, permissionsRequired: { GET: { permissions: ['view-livechat-rooms', 'view-l-room'], operation: 'hasAny' } } },
```
Comment (sampaiodiego) [inline apps/meteor/app/livechat/imports/server/rest/rooms.js:22]: you didn't remove the checks from endpoint code

File: apps/meteor/app/livechat/imports/server/rest/users.js [hunk 2]
```diff
@@ -20,41 +29,42 @@ API.v1.addRoute(
 			const { text } = this.queryParams;
 
 			if (this.urlParams.type === 'agent') {
+				if (!(await hasAtLeastOnePermission(this.userId, ['transfer-livechat-guest', 'edit-omnichannel-contact']))) {
```
Comment (sampaiodiego) [inline apps/meteor/app/livechat/imports/server/rest/users.js:32]: either `await` for the `async` version or don't use `await`
```suggestion
				if (!(await hasAtLeastOnePermissionAsync(this.userId, ['transfer-livechat-guest', 'edit-omnichannel-contact']))) {
```

File: apps/meteor/app/api/server/api.js [hunk 3]
```diff
@@ -52,6 +52,25 @@ const getRequestIP = (req) => {
 	return forwardedFor[forwardedFor.length - httpForwardedCount];
 };
 
+const checkPermisisonsForInvocation = (userId, permissionsPayload, requestMethod) => {
```
Comment (sampaiodiego) [inline apps/meteor/app/api/server/api.js:55]: (picky alert)

we might need a better definition on these, but I think we should use more named functions.. 

maybe the definition could be: use arrow functions only when creating local/scoped functions, like a function inside another function. wdyt?

File: apps/meteor/app/api/server/api.js [hunk 4]
```diff
@@ -311,21 +330,47 @@ export class APIClass extends Restivus {
 		return routeActions.map((action) => this.getFullRouteName(route, action, apiVersion));
 	}
 
+	checkPermissions(options) {
```
Comment (sampaiodiego) [inline apps/meteor/app/api/server/api.js:333]: (pick alert 2)

this method could be a function in file's scope right?

I'd also recommend not changing the `options` param, but instead having  the `permissionRequired` as the return value.

having this function in TS and having a unit test would also be good

**Review by sampaiodiego — CHANGES_REQUESTED** [review]
picky comments can be addressed in a future engineering day =)

File: apps/meteor/app/api/server/api.js (hunk 3 above)
Comment (sampaiodiego) [reply apps/meteor/app/api/server/api.js:55]: yep, I think we're missing a standard.. I'm worried about two things specifically:

- hoist variables and how that affects gc
- better stack traces

File: apps/meteor/app/livechat/imports/server/rest/users.js [hunk 5]
```diff
@@ -1,17 +1,26 @@
...
 	'livechat/users/:type',
-	{ authRequired: true },
 	{
-		get() {
+		authRequired: true,
+		permissionsRequired: {
+			GET: {
+				permissions: ['manage-livechat-agents'],
+				operation: 'hasAll',
+			},
```
Comment (murtaza98) [inline apps/meteor/app/livechat/imports/server/rest/users.js:18]: I believe there's a small issue here which changes the permission checks for this endpoint. If we take an example of agents (i.e. `:type=agent`), right now any user with **one of these 3 permission** is allowed to query this endpoint

- 'manage-livechat-agents'
- 'transfer-livechat-guest'
- 'edit-omnichannel-contact'

But with this new logic, the user will always need to have 'manage-livechat-agents' role & users with who don't have this role but do have the other 2 roles won't be able to access this endpoint

Comment (murtaza98) [inline apps/meteor/app/livechat/imports/server/rest/users.js:18]: I believe there's a small issue here which changes the permission checks for this endpoint. If we take an example of agents (i.e. `:type=agent`), right now any user with **one of these 3 permission** is allowed to query this endpoint

- 'manage-livechat-agents'
- 'transfer-livechat-guest'
- 'edit-omnichannel-contact'

But with this new logic, the user will always need to have 'manage-livechat-agents' role & users who don't have this role but do have the other 2 roles won't be able to access this endpoint

Comment (murtaza98) [reply apps/meteor/app/livechat/imports/server/rest/users.js:18]: FYI: fixed this here https://github.com/RocketChat/Rocket.Chat/pull/26552

Comment (murtaza98) [reply apps/meteor/app/livechat/imports/server/rest/users.js:18]: FYI: fixed this here https://github.com/RocketChat/Rocket.Chat/pull/26552

**Comment by murtaza98** [comment]
Hey @KevLehman This looks really good :tada: I've left just a few suggestions over here https://github.com/RocketChat/Rocket.Chat/pull/26552 - they are mostly removing some permission checks which aren't required now & addressing the issue I mentioned above

**Review by murtaza98 — DISMISSED** [review]
LGTM :tada:

**Review by sampaiodiego — APPROVED** [review]

**Review by murtaza98 — APPROVED** [review]

**Comment by KevLehman** [comment]
@sampaiodiego @murtaza98 , you'll have to re-approve this one since there's a conflict :cry:

**Comment by KevLehman** [comment]
btw, can we skip QA on this one? :eyes:

**Review by sampaiodiego — APPROVED** [review]

**Comment by sampaiodiego** [comment]
since the endpoints changed are all omnichannel, I'll let you guys decide if you want to skip QA.

**Review by murtaza98 — APPROVED** [review]

File: apps/meteor/app/api/server/api.js [hunk 6]
```diff
@@ -465,6 +462,8 @@ export class APIClass extends Restivus {
 
 						result = DDP._CurrentInvocation.withValue(invocation, () => Promise.await(originalAction.apply(this))) || API.v1.success();
 
+						console.log('------------------------------------------------------');
+						console.log(result);
```
Comment (murtaza98) [inline apps/meteor/app/api/server/api.js:466]:
```suggestion
```

**Review by murtaza98 — APPROVED** [review]

**Review by sampaiodiego — APPROVED** [review]

**Review by murtaza98 — APPROVED** [review]

**Pull request merged**